activejob/lib/active_job/locking.rb

```diff
@@ -0,0 +1,25 @@
+# frozen_string_literal: true
+
+module ActiveJob
+  # Provides behavior for controlling how many of a job can be enqueued.
+  module Locking
+    def initialize(*arguments)
+      super
+
+      @lock_timeout = self.class.lock_timeout.to_i
+      if @lock_timeout > 0 && !self.class.lock_key.nil?
+        @lock_key = self.class.lock_key.call(self)
```

Comment:
Calling the `lock_key` in the initialize may fail if you want to access the arguments in inside the proc. When a job in deserialized the arguments are not set but the initialize is called, so at this point the job may not have anything inside it. https://github.com/rails/rails/blob/master/activejob/lib/active_job/core.rb#L44

This will break the case were the job is already in the queue but you want to re-enqueue. The information to calculate the lock key will not be present when the proc is executed. I think if you write a test with a job that needs to retry you will see this problem.

Reply:
Good catch, I forgot about retries. I've reproduced this locally and am working on a good test for it and will then push that up along with a fix. My idea right now is to generate the key at the time its requested.

I did consider adding the key to the serialized data though but I'm not sure I like that. The benefit of doing that is if the proc changes while a job is enqueued it will have the same key on the retry as it did when it was first queued up. The downside is it just adds more stuff to the serialized job data.

Reply:
That is how we implemented at Shopify. The lock_key is stored in the serialized date and we only calculate it when it is requested.